Update examples

HealthyPear opened the pull request:
Examples from the README and in general from the examples folder seem to be outdated since #131.

Here I attempted to fix all of them, but for some (commented out at the moment from the list in the CI job) I could not find a suitable test file (or maybe there are deeper changes to apply to those examples). For a couple of examples I added a max-events option due to resource file truncations.

In order to ease the user experience, and also for the CI to not worry about specific dependencies, I created anew extra called 'examples'.
I noticed that when using python 3.13 something is pulling numba > 0.56 (either astropy or ctapipe) and this makes the installation crash. I tried different numba versions, but in the end limiting python to 3.12 proved to be the more most stable solution.

P.S: there is no documentation for this package, so I do not know if it makes sense, but you couldadd jupytext and make the examples folder a [sphinx-gallery](https://sphinx-gallery.github.io/stable/index.html)

maxnoe commented:
> I noticed that when using python 3.13 something is pulling numba > 0.56 (either astropy or ctapipe) and this makes the installation crash. 

3.13 is not yet supported by numba. But why is numba a topic here?

maxnoe commented:
> But why is numba a topic here?

Ah, I see, one of the examples i using ctapipe. However, this example is anyways terribly outdated. I'd just remove it

HealthyPear commented:
> > I noticed that when using python 3.13 something is pulling numba > 0.56 (either astropy or ctapipe) and this makes the installation crash.
> 
> 3.13 is not yet supported by numba. But why is numba a topic here?

some examples use ctapipe and I guess this is calling numba?

HealthyPear commented:
there is a CI failure that I did not get locally (I use https://nektosact.com/) I'll try to see what it the problem

maxnoe commented on .github/workflows/ci.yml:
```diff
@@ -43,4 +43,23 @@ jobs:
         run: |
           pytest --cov=eventio --cov-report=xml
 
+      - name: Run examples
```
probably should set `MPLBACKEND=agg` to not create windows in the CI for the plots

HealthyPear replied:
I do not see a change in adding it, but no strong feelings

HealthyPear replied:
done

maxnoe commented:
cibuildwheel tries building for python 3.6... that's not been supported for a very long time now.

maxnoe commented on .github/workflows/ci.yml:
```diff
@@ -19,7 +19,7 @@ jobs:
     strategy:
       matrix:
-        python-version: ["3.10", "3.11", "3.12", "3.13"]
+        python-version: ["3.10", "3.11", "3.12"]
```
Please do not remove testing on 3.13

HealthyPear replied:
reverted

maxnoe commented on pyproject.toml:
```diff
@@ -6,7 +6,7 @@ build-backend = "setuptools.build_meta"
 description = "Python read-only implementation of the EventIO file format"
 
-requires-python = '>=3.10'
+requires-python = '<3.13'
```
no, we are compatible with 3.13. And we do require at least 3.10.

This is why the CI is failing to try to build wheels for 3.6

HealthyPear replied:
reverted

HealthyPear commented:
> otherwise we should make a new job only for the examples in which we make a new matrix without 3.13

this works if you want

HealthyPear commented:
OK I added a separate job only for examples where I removed python 3.13.

I you confirm that you want to remove all 3 examples that use ctapipe I can re-merge the 2 jobs into 1

maxnoe commented on .github/workflows/ci.yml:
```diff
@@ -44,3 +44,42 @@ jobs:
+
+          MPLBACKEND=agg
+          
+          # python iter_mc_events.py
```
Why are these commented out? These should work on any simtel test file  in the resources

HealthyPear replied:
I could not make them work unfortunately

some give StopIteration error, some EOF Error, etc

this one specifically for example gives me

```python
➜ python iter_mc_events.py ../tests/resources/gamma_100gev_1216.eventio
/Users/michele/Software/pyeventio/src/eventio/simtel/simtelfile.py:251: UnknownObjectWarning: object type encountered, which is no handled at the moment: RunHeader[1200]
  warnings.warn(
/Users/michele/Software/pyeventio/src/eventio/simtel/simtelfile.py:251: UnknownObjectWarning: object type encountered, which is no handled at the moment: TelescopeDefinition[1201]
  warnings.warn(
/Users/michele/Software/pyeventio/src/eventio/simtel/simtelfile.py:251: UnknownObjectWarning: object type encountered, which is no handled at the moment: EventHeader[1202](event_id=1)
  warnings.warn(
/Users/michele/Software/pyeventio/src/eventio/simtel/simtelfile.py:251: UnknownObjectWarning: object type encountered, which is no handled at the moment: ArrayOffsets[1203]
  warnings.warn(
/Users/michele/Software/pyeventio/src/eventio/simtel/simtelfile.py:251: UnknownObjectWarning: object type encountered, which is no handled at the moment: EventEnd[1209](event_id=1)
  warnings.warn(
/Users/michele/Software/pyeventio/src/eventio/simtel/simtelfile.py:251: UnknownObjectWarning: object type encountered, which is no handled at the moment: RunEnd[1210]
  warnings.warn(
Traceback (most recent call last):
  File "/Users/michele/Software/pyeventio/examples/iter_mc_events.py", line 17, in <module>
    with SimTelFile(args.inputfile) as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/michele/Software/pyeventio/src/eventio/simtel/simtelfile.py", line 128, in __init__
    self.next_low_level()
  File "/Users/michele/Software/pyeventio/src/eventio/simtel/simtelfile.py", line 150, in next_low_level
    o = next(self)
        ^^^^^^^^^^
  File "/Users/michele/Software/pyeventio/src/eventio/base.py", line 115, in __next__
    read_sync_marker(self)
  File "/Users/michele/Software/pyeventio/src/eventio/base.py", line 183, in read_sync_marker
    check_size_or_raise(sync, constants.SYNC_MARKER_SIZE, zero_ok=True)
  File "/Users/michele/Software/pyeventio/src/eventio/base.py", line 166, in check_size_or_raise
    raise StopIteration
StopIteration
```

HealthyPear replied:
ah with camsettings_v6.simtel.zst seems to work

HealthyPear replied:
`plot_light_on_ground_by_particle.py` and `plot_light_on_ground_by_particle.py` tell me there is no `emitter` in files similar to the one I use for `plot_light_on_ground.py`

```python
➜ python examples/plot_light_on_ground_by_particle.py tests/resources/3_gammas_reuse_5.dat
Traceback (most recent call last):
  File "/Users/michele/Software/pyeventio/examples/plot_light_on_ground_by_particle.py", line 113, in <module>
    main()
  File "/Users/michele/Software/pyeventio/examples/plot_light_on_ground_by_particle.py", line 44, in main
    mask = np.isclose(emitter['mass'], mass)
                      ~~~~~~~^^^^^^^^
TypeError: 'NoneType' object is not subscriptable
```

`read_all_objects.py` instead returns this with many different files

```python
Traceback (most recent call last):
  File "/Users/michele/Software/pyeventio/examples/read_all_objects.py", line 16, in <module>
    o.parse()
  File "/Users/michele/Software/pyeventio/src/eventio/base.py", line 346, in parse
    raise ValueError('This object only contains subobjects')
ValueError: This object only contains subobjects
```

maxnoe replied:
> tell me there is no emitter in files 

We indeed have no file with `emitter` information. I'll produce one.

HealthyPear replied:
I think the only one missing is read_all_objects.py which gives me the same error no matter the test file
